opendevin/runtime/plugins/agent_skills/agentskills.py

```diff
@@ -508,34 +510,122 @@ def _edit_or_append_file(
 
 
 @update_pwd_decorator
-def edit_file(file_name: str, start: int, end: int, content: str) -> None:
-    """Edit a file.
+def edit_file(file_name: str, to_replace: str, new_content: str) -> None:
```

Comment:
> Imo this raises the chances of an accidental replacement at the wrong occurence as this has to work with any kind of text, not just with files from eval data, that might be "unique".

Definitely a possibility - maybe it is just a question of "how much". 

Ideally, I very much like the line-based editing approach because it is clean, complete, and doesn't involve too much heuristics.

But these line-based edits are mainly bottlenecked by the current LLM's capability - things may improve in a year, but in the meanwhile, aside from the training model (which we are already thinking & doing), exploring different editing approaches like this could be another good alternative in improving the agent quality.

Reply:
I'm curious about your screenshot: did the LLM edit that same file already before?
That could explain that it itself doesn't keep track of line numbers and didn't care to re-read the changed file before applying an edit "blindly" the 2nd time.
This could be avoided by a slightly changed prompt, potentially, by advising the LLM to actually refresh its context about the location _before_ it wants to edit again.
Even in my use of LLM's as assistants in my IDE (Cursor), I have to double-check every code suggestion as it constantly makes assumptions and with only a small "task" in mind, looses the big picture, and ignores consequences from related issues it may cause... but I digress. 😂

Reply:
> did the LLM edit that same file already before?

Yes it did!

> This could be avoided by a slightly changed prompt, potentially, by advising the LLM to actually refresh its context about the location before it wants to edit again.

Hmm, I think this is also possible! But still, the other possibility is that this is too challenging for today's LLM to figure out "line number" - Let's wait for score haha 😂

Reply:
It's more from what I experienced in using OpenDevin where there were these cases, as you observed, where it made multiple edits _without_ checking first. 
I'd _hope_ LLM's would get trained to act accordingly, but for now I'd just make more precise instructions for it to follow.